## custom inputs usage


gotcha can also be also operated using custom inputs, consisting of either 


-  ```.fna``` fasta-formatted set of unaligned sequences, with the ```-cf``` flag
-  ```.aln``` fasta-formatted set of aligned sequences, with the ```-ca``` flag
-  ```.nwk``` newik-formatted phylogenetic tree, with the ```-cn``` flag


--- 


Using a custom set of sequences, gotcha will perform all the steps, excluding the taxonomic assignation filetering. Here is an example
of using the ```-cf``` flag

```python gotcha.py --marker coding -e -v -cf custom.fna -o custom_baits -c 5```


--- 


Using a custom alignment, gotcha will skip all preprocessing step and go directly to size selection, tree inference, ancestral state reconstruction and baits desing:
This can be achieved using the ```-ca``` flag:

```python gotcha.py --marker coding -e -v -ca custom.aln -o custom_baits -c 5```


--- 


gotcha also supports custom phylogenies, wich can be used alongside either a custom set of aligned or unaligned sequences. The tree can lack some of the species - which will be appended to the tree. 
This can be achieved using the ```-cn``` flag:


```python gotcha.py --marker coding -e -v -ca custom.aln -ct custom.nwk -o custom_baits -c 5```


--- 


[back to main](https://github.com/Kevinnota/gotcha/blob/main/0.md)